packages/storage-plus/src/keys.rs

```diff
@@ -225,18 +255,47 @@ impl<'a> PrimaryKey<'a> for Addr {
     type Suffix = Self;
     type SuperSuffix = Self;
 
-    fn key(&self) -> Vec<&[u8]> {
+    fn key(&self) -> Vec<Key> {
         // this is simple, we don't add more prefixes
-        vec![self.as_bytes()]
+        vec![Key::Ref(self.as_bytes())]
     }
 }
 
 impl<'a> Prefixer<'a> for Addr {
-    fn prefix(&self) -> Vec<&[u8]> {
-        vec![&self.as_bytes()]
+    fn prefix(&self) -> Vec<Key> {
+        vec![Key::Ref(self.as_bytes())]
     }
 }
 
+macro_rules! integer_key {
+    (for $($t:ty, $v:tt),+) => {
+        $(impl<'a> PrimaryKey<'a> for $t {
+            type Prefix = ();
+            type SubPrefix = ();
+            type Suffix = Self;
+            type SuperSuffix = Self;
+
+            fn key(&self) -> Vec<Key> {
+                vec![Key::$v(self.to_be_bytes())]
+            }
+        })*
+    }
+}
+
+integer_key!(for i8, Val8, u8, Val8, i16, Val16, u16, Val16, i32, Val32, u32, Val32, i64, Val64, u64, Val64);
```

Comment:
Huh? I am a bit lost here with the macros, wouldn't it be:

`integer_key!(i8, Val8);`
`integer_key!(u8, Val8);`
`integer_key!(i16, Val16);`
`integer_key!(u16, Val16);`
...

Or that `for i8` does something magic?

Reply:
It's the `for` keyword yes. Sadly, with declarative macros we cannot take the type (`i8`), obtain the size (`8`), and build an identifier (`Val8`). That's why those are repeated. I've heard something like that can be done with procedural macros, though.